pkg/source/source.go

```diff
@@ -64,200 +60,68 @@ func Kind[T client.Object](cache cache.Cache, object T, handler handler.TypedEve
 func WithPredicates[T any](p ...predicate.TypedPredicate[T]) ChannelOpt[T] {
-	return func(c *channel[T]) {
+	return func(c *channelOpts[T]) {
 		c.predicates = append(c.predicates, p...)
 	}
 }
 
 // WithBufferSize configures the buffer size for a source.Channel. By
 // default, the buffer size is 1024.
 func WithBufferSize[T any](bufferSize int) ChannelOpt[T] {
-	return func(c *channel[T]) {
-		c.bufferSize = &bufferSize
+	return func(c *channelOpts[T]) {
+		c.bufferSize = bufferSize
+	}
+}
+
+// NewChannelBroadcaster creates a new ChannelBroadcaster for the given channel.
+// A ChannelBroadcaster is a wrapper around a channel that allows multiple listeners to all
+// receive the events from the channel.
+func NewChannelBroadcaster[T any](source <-chan event.TypedGenericEvent[T]) *internal.ChannelBroadcaster[T] {
+	return &internal.ChannelBroadcaster[T]{
+		Source: source,
 	}
 }
 
 // Channel is used to provide a source of events originating outside the cluster
 // (e.g. GitHub Webhook callback).  Channel requires the user to wire the external
 // source (e.g. http handler) to write GenericEvents to the underlying channel.
-func Channel[T any](source <-chan event.TypedGenericEvent[T], handler handler.TypedEventHandler[T], opts ...ChannelOpt[T]) Source {
-	c := &channel[T]{
-		source:  source,
-		handler: handler,
+func Channel[T any](broadcaster *internal.ChannelBroadcaster[T], handler handler.TypedEventHandler[T], opts ...ChannelOpt[T]) Source {
+	c := &channelOpts[T]{
+		bufferSize: 1024,
 	}
 	for _, opt := range opts {
 		opt(c)
 	}
 
-	return c
-}
-
-type channel[T any] struct {
-	// once ensures the event distribution goroutine will be performed only once
-	once sync.Once
-
-	// source is the source channel to fetch GenericEvents
-	source <-chan event.TypedGenericEvent[T]
-
-	handler handler.TypedEventHandler[T]
-
-	predicates []predicate.TypedPredicate[T]
-
-	bufferSize *int
-
-	// dest is the destination channels of the added event handlers
-	dest []chan event.TypedGenericEvent[T]
-
-	// destLock is to ensure the destination channels are safely added/removed
-	destLock sync.Mutex
-}
-
-func (cs *channel[T]) String() string {
-	return fmt.Sprintf("channel source: %p", cs)
-}
-
-// Start implements Source and should only be called by the Controller.
-func (cs *channel[T]) Start(
-	ctx context.Context,
-	queue workqueue.RateLimitingInterface,
-) error {
-	// Source should have been specified by the user.
-	if cs.source == nil {
-		return fmt.Errorf("must specify Channel.Source")
-	}
-	if cs.handler == nil {
-		return errors.New("must specify Channel.Handler")
-	}
-
-	if cs.bufferSize == nil {
-		cs.bufferSize = ptr.To(1024)
-	}
-
-	dst := make(chan event.TypedGenericEvent[T], *cs.bufferSize)
-
-	cs.destLock.Lock()
-	cs.dest = append(cs.dest, dst)
-	cs.destLock.Unlock()
-
-	cs.once.Do(func() {
-		// Distribute GenericEvents to all EventHandler / Queue pairs Watching this source
-		go cs.syncLoop(ctx)
```

Comment:
This is the main issue that this PR solves: here, the context passed to the first call of Start is used to manage the goroutine that is used for all future calls to Start.